File: src/ipc/messages/types/structs/Some.ts
import * as ldm from '../../leapdatamodel';
import * as valuetype from '../../valuetype';
import type * as types from '..';
import { StructBase as _StructBase } from '../../StructBase';
import { LeapError as _LeapError } from '../../LeapError';
import { ValueBase as _ValueBase } from '../../ValueBase';

export class Some<T> extends _StructBase {
    static _type: string = 'some';
    static _leapStruct: valuetype.LeapStruct = new valuetype.LeapStruct("some", ["t"], [new valuetype.Property("value", new valuetype.ValueType("t", []))]);
    value: T;

    constructor(value: T, ) {
        super();
        this.value = value;
    }

    _verifyAsTypeOrRaise(path: string[], valueType: valuetype.ValueType): void {
        if (valueType.name == Some._type) {
            const structType = Some._leapStruct.applyArgs(valueType.args);
            _ValueBase._verifyValueOrRaise(path.concat(['value']), this.value, structType.props[0].propType);
        } else {
            throw _ValueBase._pathError(path, `expecting ${Some._type}`);
        }
    }

    _toLdm(valueType: valuetype.ValueType): ldm.Value {
        let value = new Map<string, ldm.Value>();
        const structType = Some._leapStruct.applyArgs(valueType.args);
        value.set("value", _ValueBase._valueToLdm(this.value, structType.props[0].propType));
        return new ldm.StructValue(value, valueType);
    }
}

File: src-tauri/src/ipc/messages/types/structs/some.rs
use std::collections::HashMap as _HashMap;
#[allow(unused_imports)]
use super::super::super::types;
use super::super::super::value_base::ValueBase as _ValueBase;
use super::super::super::struct_base::StructBase as _StructBase;
use super::super::super::leap_data_model as ldm;
use super::super::super::value_type::ValueType as _ValueType;

#[derive(Clone, Debug)]
pub struct Some<T: _ValueBase> {
    pub value: T,
}

impl<T: _ValueBase> Some<T> {
    pub fn new(value: T, ) -> Self {
        Self {
            value,
        }
    }
}

impl<T: _ValueBase> _ValueBase for Some<T> {
    const _TYPE: &'static str = "some";

    fn _applied_args() -> Vec<_ValueType> {
        vec![
            _ValueType::new(T::_TYPE.to_owned(), T::_applied_args()),
        ]
    }

    fn _to_ldm(&self) -> ldm::Value {
        ldm::Value::Struct(self._to_ldm_struct())
    }
}

impl<T: _ValueBase> _StructBase for Some<T> {
    fn _to_ldm_struct(&self) -> ldm::StructValue {
        let mut value = _HashMap::new();
        value.insert("value".to_owned(), self.value._to_ldm());
        ldm::StructValue{
            value,
            value_type: _ValueType::new(Self::_TYPE.to_owned(), Self::_applied_args()),
        }
    }
}